core/src/main/java/cn/edu/tsinghua/iot/benchmark/entity/Batch/MultiDeviceBatch.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package cn.edu.tsinghua.iot.benchmark.entity.Batch;
+
+import cn.edu.tsinghua.iot.benchmark.entity.Record;
+import cn.edu.tsinghua.iot.benchmark.schema.schemaImpl.DeviceSchema;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+public class MultiDeviceBatch implements IBatch {
+
+  private final ArrayList<DeviceSchema> deviceSchemas;
+  private final ArrayList<List<Record>> recordLists;
+
+  private int index = 0;
+
+  public MultiDeviceBatch(int size) {
+    this.deviceSchemas = new ArrayList<>(size);
+    this.recordLists = new ArrayList<>(size);
+    for (int i = 0; i < size; i++) {
+      deviceSchemas.add(null);
+      recordLists.add(new LinkedList<>());
```

Comment:
If we give the size of deviceSchemas when init, maybe there we are no need to add null into deviceSchemas?

Reply:
Without `add(null)`, `deviceSchemas.set(index, deviceSchema)` will cause OutOfIndexException.
I think `new ArrayList<>(size)` only means pre-allocate of memory.

Reply:
please take a look at [this commit](https://github.com/thulab/iot-benchmark/pull/338/commits/880dfbdf74b67a2fe1a786d5d9c32df2e09813be), i think the problem has been solved